src/minerva_lib/crop.py

```diff
@@ -0,0 +1,298 @@
+
+    Arguments:
+        input_shape: The width, height at pyramid level 0
+        level_count: Number of available pyramid levels
+        max_size: Maximum output image extent in x or y
+        round_up: Use higher-magnifcation pyramid level contained by
+                  `max_size` if true. Use lower-magnification pyramid
+                  level containing `max_size` if false.
+
+    Returns:
+        Integer power of 2 pyramid level
+    '''
+
+    if max_size is None:
+        return 0
+
+    longest_side = max(*input_shape)
+    ratio_log = np.log2(longest_side / max_size)
+
+    if round_up:
+        level = np.ceil(ratio_log)
+    else:
+        level = np.floor(ratio_log)
+
+    return int(np.clip(level, 0, level_count - 1))
+
+
+def scale_by_pyramid_level(coordinates, level):
+    ''' Returns the coordinates measured at a given pyramid level.
+
+    Arguments:
+        coordrnates: Coordinates to downscale by _level_
+        level: Integer power of 2 pyramid level
+
+    Returns:
+        Downscaled integer coordinates
+    '''
+
+    scaled_coords = np.array(coordinates) / (2 ** level)
+    return np.int64(np.floor(scaled_coords))
+
+
+def get_tile_start(tile_size, crop_origin):
+    ''' Returns the indices of the first tile in the region.
+
+    Args:
+        tile_size: width, height of one tile
+        crop_origin: x, y origin of requested image region
+        crop_size: width, height of requested image region
+
+    Returns:
+        Two-item int64 array i, j tile indices
+    '''
+    start = np.array(crop_origin)
+
+    return np.int64(np.floor(start / tile_size))
+
+
+def get_tile_count(tile_size, crop_origin, crop_size):
+    ''' Returns the number of tiles in the region.
+
+    Args:
+        tile_size: width, height of one tile
+        crop_origin: x, y origin of requested image region
+        crop_size: width, height of requested image region
+        image_size: width, height of full image
+
+    Returns:
+        Two-item int64 array i, j tile indices
+    '''
+    end = np.array(crop_origin) + crop_size
+
+    return np.int64(np.ceil(end / tile_size))
+
+
+def get_subregion(indices, tile_size, crop_origin, crop_size):
+    '''Determines the region within this specific tile
+    that is required for the requested image region.
+
+    Args:
+        indices: integer i, j tile indices
+        tile_size: width, height of one tile
+        crop_origin: x, y origin of requested image region
+        crop_size: width, height of requested image region
+
+    Returns:
+        The part of the tile within the requested region
+        Subregion start, end relative to origin of tile
+    '''
+
+    tile_start = np.int64(indices) * tile_size
+    crop_end = np.int64(crop_origin) + crop_size
+    tile_end = tile_start + tile_size
+
+    # Should be true that 0 <= start_uv < end_uv <= tile_size
+    start_uv = np.maximum(crop_origin, tile_start) - tile_start
```

Comment:
In x, or y at starts at the start of the tile or the start of the requested region.